pep-0678.rst

```diff
@@ -0,0 +1,267 @@
+
+Existing approaches must pass this additional information around while keeping
+it in sync with the state of raised, and potentially caught or chained, exceptions.
+This is already error-prone, and made more difficult by :pep:`654` ``ExceptionGroup`` s,
+so the time is right for a built-in solution.  We therefore propose to add a mutable
+field ``__note__`` to ``BaseException`` , which can be assigned a string - and
+if assigned, is automatically displayed in formatted tracebacks.
+
+
+Example usage
+-------------
+
+   >>> try:
+   ...     raise TypeError('bad type')
+   ... except Exception as e:
+   ...     e.__note__ = 'Add some information'
+   ...     raise
+   ...
+   Traceback (most recent call last):
+     File "<stdin>", line 2, in <module>
+   TypeError: bad type
+   Add some information
+   >>>
+
+When collecting exceptions into an exception group, we may want
+to add context information for the individual errors. In the following
+example with `Hypothesis' proposed support for ExceptionGroup
+<https://github.com/HypothesisWorks/hypothesis/pull/3191>`__, each
+exception includes a note of the minimal failing example::
+
+    from hypothesis import given, strategies as st, target
+
+    @given(st.integers())
+    def test(x):
+        assert x < 0
+        assert x > 0
+
+
+    + Exception Group Traceback (most recent call last):
+    |   File "test.py", line 4, in test
+    |     def test(x):
+    |
+    |   File "hypothesis/core.py", line 1202, in wrapped_test
+    |     raise the_error_hypothesis_found
+    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+    | ExceptionGroup: Hypothesis found 2 distinct failures.
+    +-+---------------- 1 ----------------
+        | Traceback (most recent call last):
+        |   File "test.py", line 6, in test
+        |     assert x > 0
+        |     ^^^^^^^^^^^^
+        | AssertionError: assert -1 > 0
+        |
+        | Falsifying example: test(
+        |     x=-1,
+        | )
+        +---------------- 2 ----------------
+        | Traceback (most recent call last):
+        |   File "test.py", line 5, in test
+        |     assert x < 0
+        |     ^^^^^^^^^^^^
+        | AssertionError: assert 0 < 0
+        |
+        | Falsifying example: test(
+        |     x=0,
+        | )
+        +------------------------------------
+
+
+Specification
+=============
+
+``BaseException`` gains a new mutable attribute ``__note__`` , which defaults to
+``None`` and may have a string assigned.  When an exception with a note is displayed,
+the note is displayed immediately after the exception.
+
+Assigning a new string value overrides an existing note; if concatenation is desired
+users are responsible for implementing it with e.g.::
+
+    e.__note__ = msg if e.__note__ is None else e.__note__ + "\n" + msg
+
+It is an error to assign a non-string-or-``None`` value to ``__note__`` ,
+or to attempt to delete the attribute.
+
```

Comment:
:meth:`ExceptionGroup.subgroup` and :meth:`ExceptionGroup.split` copy the ``__note__`` of the original exception group to the parts.